src/arextract.py

```diff
@@ -106,6 +109,209 @@ def boxcar(slf, det, specobjs, sciframe, varframe, skyframe, crmask, scitrace):
+            msgs.warn("Low extracted flux for obj={:s}.  Not ready for Optimal".format(specobjs[o].idx))
+            scitrace['opt_profile'].append({})
+            continue
+    # QA
+    arqa.obj_profile_qa(slf, specobjs, scitrace)
+    return scitrace['opt_profile']
+
+
+def optimal_extract(slf, det, specobjs, sciframe, varframe, skyframe, crmask, scitrace,
+                 pickle_file=None, profiles=None):
+    """ Preform optimal extraction
+    Standard Horne approach
+    Parameters
+    ----------
+    slf
+    det
+    specobjs
+    sciframe
+    varframe
+    crmask
+    scitrace
+    COUNT_LIM
+    pickle_file
+    Returns
+    -------
+    """
+    '''
+    import pickle
+    if pickle_file is not None:
+        f = open(pickle_file,'r')
+        args = pickle.load(f)
+        f.close()
+        det, specobjs, sciframe, varframe, skyframe, crmask, scitrace, tilts = args
+        slf = None
+        scitrace['opt_profile'] = profiles
+    else:
+        tilts = None
+    '''
+    # Setup
+    model_var = np.abs(skyframe + sciframe - np.sqrt(2)*slf._msrn[det-1]) + \
+                 slf._msrn[det-1]**2  # sqrt 2 term deals with negative flux/sky
+    model_ivar = 1./model_var
+    msgs.work("Consider making a model of the object for model_ivar")
+    cr_mask = 1.0-crmask
+    # Loop
+    nobj = scitrace['traces'].shape[1]
+    for o in range(nobj):
+        # Fit dict
+        fit_dict = scitrace['opt_profile'][o]
+        if 'param' not in fit_dict.keys():
+            continue
+        # Slit image
+        slit_img = artrace.slit_image(slf, det, scitrace, o)#, tilts=tilts)
+        #msgs.warn("Turn off tilts")
+        # Object pixels
+        weight = scitrace['object'][:,:,o]
+        gdo = (weight > 0) & (model_ivar > 0)
+        # Profile image
+        prof_img = np.zeros_like(weight)
+        prof_img[gdo] = arutils.func_val(fit_dict['param'], slit_img[gdo],
+                                         fit_dict['func'])
+        # Normalize
+        norm_prof = np.sum(prof_img, axis=1)
+        prof_img /= np.outer(norm_prof + (norm_prof == 0.), np.ones(prof_img.shape[1]))
+        # Mask (1=good)
+        mask = np.zeros_like(prof_img)
+        mask[gdo] = 1.
+        mask *= cr_mask
+
+        # Optimal flux
+        opt_num = np.sum(mask * sciframe * model_ivar * prof_img, axis=1)
+        opt_den = np.sum(mask * model_ivar * prof_img**2, axis=1)
+        opt_flux = opt_num / (opt_den + (opt_den == 0.))
+        # Optimal wave
+        opt_num = np.sum(mask * slf._mswave[det-1] * model_ivar * prof_img**2, axis=1)
+        opt_wave = opt_num / (opt_den + (opt_den == 0.))
+        # Optimal ivar
+        opt_num = np.sum(mask * model_ivar * prof_img**2, axis=1)
+        ivar_den = np.sum(mask * prof_img, axis=1)
+        opt_ivar = opt_num / (ivar_den + (ivar_den==0.))
+
+        # Save
+        specobjs[o].optimal['wave'] = opt_wave*u.AA  # Yes, units enter here
+        specobjs[o].optimal['counts'] = opt_flux
```

Comment:
Should these arrays be copies? (i.e. `opt_flux.copy()` etc.). If nobj is greater than 1, I worry that the stored arrays will reference the updated opt_flux array (but I'm not sure). It might be best to quickly check this. (i.e. after this line, multiply opt_flux by 0.0 (i.e. `opt_flux *= 0.0`, and then print `specobjs[o].optimal['counts']` to make sure it's not full of zeros).